Source/WebCore/rendering/svg/SVGTextQuery.cpp

```diff
@@ -137,63 +137,53 @@ bool SVGTextQuery::executeQuery(Data* queryData, ProcessTextFragmentCallback fra
 
 bool SVGTextQuery::mapStartEndPositionsIntoFragmentCoordinates(Data* queryData, const SVGTextFragment& fragment, unsigned& startPosition, unsigned& endPosition) const
 {
-    // Reuse the same logic used for text selection & painting, to map our query start/length into start/endPositions of the current text fragment.
+    // Make <startPosition, endPosition> offsets relative to the current text box.
     ASSERT(startPosition >= queryData->processedCharacters);
     ASSERT(endPosition >= queryData->processedCharacters);
-    startPosition -= queryData->processedCharacters;
-    endPosition -= queryData->processedCharacters;
 
-    // <startPosition, endPosition> is now a tuple of offsets relative to the current text box.
-    // Compute the offsets of the fragment in the same offset space.
-    unsigned fragmentStartInBox = fragment.characterOffset - queryData->textBox->start();
-    unsigned fragmentEndInBox = fragmentStartInBox + fragment.length;
+    unsigned boxStart = queryData->processedCharacters;
 
-    // Check if the ranges intersect.
-    startPosition = std::max<unsigned>(startPosition, fragmentStartInBox);
-    endPosition = std::min<unsigned>(endPosition, fragmentEndInBox);
+    startPosition -= boxStart;
+    endPosition -= boxStart;
 
-    if (startPosition >= endPosition)
-        return false;
-
-    modifyStartEndPositionsRespectingLigatures(queryData, fragment, startPosition, endPosition);
+    // Reuse the same logic used for text selection & painting, to map our
+    // query start/length into start/endPositions of the current text fragment.
     if (!queryData->textBox->mapStartEndPositionsIntoFragmentCoordinates(fragment, startPosition, endPosition))
         return false;
 
+    modifyStartEndPositionsRespectingLigatures(queryData, fragment, startPosition, endPosition);
     ASSERT_WITH_SECURITY_IMPLICATION(startPosition < endPosition);
     return true;
 }
 
 void SVGTextQuery::modifyStartEndPositionsRespectingLigatures(Data* queryData, const SVGTextFragment& fragment, unsigned& startPosition, unsigned& endPosition) const
 {
-    SVGTextLayoutAttributes* layoutAttributes = queryData->textRenderer->layoutAttributes();
-    Vector<SVGTextMetrics>& textMetricsValues = layoutAttributes->textMetricsValues();
+    Vector<SVGTextMetrics>& textMetricsValues = queryData->textRenderer->layoutAttributes()->textMetricsValues();
 
     unsigned textMetricsOffset = fragment.metricsListOffset;
+    unsigned fragmentOffset = 0;
+    unsigned fragmentEnd = fragment.length;
```

Comment:
Now `startPosition` and  `endPosition` are both in _Fragment coordinates_. But I do not understand why this is simplified than the original code.

I think it is clearer to have `SVGInlineTextBox` and `SVGTextFragment` in _Line coordinates_ and have the calculations of `startPosition` and `endPosition` in _InlineTextBox coordinates_.  The conversion to the `Fragment coordinates` and intersecting them with the Fragment boundaries should the last step. But to run some calculations in _InlineTextBox coordinates_ and others in _Fragment coordinates_ is confusing.